common/persistence/nosql/nosqlplugin/cassandra/gocql/errors.go

```diff
@@ -39,33 +41,49 @@ func ConvertError(
 	operation string,
 	err error,
 ) error {
-
-	switch err {
-	case nil:
+	if err == nil {
 		return nil
-	case context.DeadlineExceeded, gocql.ErrTimeoutNoResponse, gocql.ErrConnectionClosed:
+	}
+
+	if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, gocql.ErrTimeoutNoResponse) || errors.Is(err, gocql.ErrConnectionClosed) {
 		return &persistence.TimeoutError{Msg: fmt.Sprintf("operation %v encountered %v", operation, err.Error())}
-	case gocql.ErrNotFound:
+	}
+	if errors.Is(err, gocql.ErrNotFound) {
 		return serviceerror.NewNotFound(fmt.Sprintf("operation %v encountered %v", operation, err.Error()))
 	}
 
-	switch v := err.(type) {
-	case *gocql.RequestErrWriteTimeout:
-		return &persistence.TimeoutError{Msg: fmt.Sprintf("operation %v encountered %v", operation, err.Error())}
-	case gocql.RequestError:
-		if v.Code() == gocql.ErrCodeOverloaded {
+	var cqlTimeoutErr gocql.RequestErrWriteTimeout
+	if errors.As(err, &cqlTimeoutErr) {
+		return &persistence.TimeoutError{Msg: fmt.Sprintf("operation %v encountered %v", operation, cqlTimeoutErr.Error())}
+	}
+
+	var cqlRequestErr gocql.RequestError
+	if errors.As(err, &cqlRequestErr) {
+		if cqlRequestErr.Code() == gocql.ErrCodeOverloaded {
 			return &serviceerror.ResourceExhausted{
 				Cause:   enumspb.RESOURCE_EXHAUSTED_CAUSE_SYSTEM_OVERLOADED,
 				Scope:   enumspb.RESOURCE_EXHAUSTED_SCOPE_SYSTEM,
-				Message: fmt.Sprintf("operation %v encountered %v", operation, err.Error()),
+				Message: fmt.Sprintf("operation %v encountered %v", operation, cqlRequestErr.Error()),
+			}
+		}
+
+		if cqlRequestErr.Code() == gocql.ErrCodeInvalid {
+			if strings.Contains(strings.ToLower(cqlRequestErr.Message()), "disk usage exceeds failure threshold") {
+				return &serviceerror.ResourceExhausted{
+					// TODO(alextalks): correct to PERSISTENCE_STORAGE_LIMIT once https://github.com/temporalio/api/pull/417 merges.
+					Cause:   enumspb.RESOURCE_EXHAUSTED_CAUSE_PERSISTENCE_LIMIT,
+					Scope:   enumspb.RESOURCE_EXHAUSTED_SCOPE_SYSTEM,
+					Message: fmt.Sprintf("operation %v encountered %v", operation, cqlRequestErr.Error()),
+				}
 			}
+
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("operation %v encountered %v", operation, cqlRequestErr.Error()))
```

Comment:
This changes the impact of this error on our SLA and needs to be reviewed.

What's the impetus behind changing it to InvalidArgument from Unavailable?

Reply:
Does it?  I am not sure I understand the SLA impact but the reason for changing it is because Cassandra's [`0x2200` error code is intended for "invalid arguments"](https://github.com/apache/cassandra/blob/7337fc0ab42a14623c96f258c1dfd19beaab54b6/doc/native_protocol_v5.spec#L1402).

If it's better to leave as is, that's fine, but it seemed incorrect to me.

Reply:
IIRC InvalidArgument doesn't affect our SLA as we return that when a user of temporal supplies an invalid argument. At this layer the problem is temporal-caused so should impact our SLAs. @yiminc thoughts?

Reply:
Ah got it.  Makes sense, I changed back to `Unavailable`.